python/hail/linalg/blockmatrix.py

```diff
@@ -51,23 +51,100 @@ def read(cls, path):
+
+        Examples
+        --------
+
+        >>> import numpy as np
+        >>> a = np.random.rand(10, 20)
+        >>> a.tofile('/local/file') # doctest: +SKIP
+
+        To create a block matrix of the same dimensions:
+
+        >>> from hail.linalg import BlockMatrix
+        >>> bm = BlockMatrix.fromfile('file:///local/file', 10, 20) # doctest: +SKIP
+
+        See also :meth:`BlockMatrix.from_numpy`.
+
+        Notes
+        -----
+        This method, analogous to `numpy.fromfile
+        <https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.fromfile.html>`__,
+        reads a binary file of float64 values in row-major order, such as that
+        produced by `numpy.tofile
+        <https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.fromfile.html>`__
+        or :meth:`BlockMatrix.tofile`.
+
+        The number of entries must be less than :math:`2^{31}`.
+
+        Use :meth:`BlockMatrix.write` and :meth:`BlockMatrix.read` to save and
+        restore block matrices.
+
+        Parameters
+        ----------
+        uri: :obj:`str`, optional
+            URI of binary input file.
+        n_rows: :obj:`int`
+            Number of rows.
+        n_cols: :obj:`int`
+            Number of columns.
+        block_size: :obj:`int`, optional
+            Block size. Default given by :meth:`default_block_size`.
+        """
+        if not block_size:
+            block_size = BlockMatrix.default_block_size()
+
+        n_entries = n_rows * n_cols
+        if n_entries >= 2 << 31:
+            raise FatalError('Number of entries must be less than 2^31, found {}'.format(n_entries))
+
+        hc = Env.hc()
+        bdm = Env.hail().utils.richUtils.RichDenseMatrixDouble.readDoubles(hc._jhc, uri, n_rows, n_cols, True)
+
+        return cls(Env.hail().linalg.BlockMatrix.fromBreezeMatrix(hc._jsc, bdm, block_size))
+
     @classmethod
     @typecheck_method(ndarray=np.ndarray,
                       block_size=nullable(int))
     def from_numpy(cls, ndarray, block_size=None):
-        """Distribute a `NumPy ndarray
+        """Distributes a `NumPy ndarray
         <https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html>`__
         as a block matrix.
 
-        Warning
-        -------
-        This method is not efficient for large matrices.
+        Examples
+        --------
+
+        >>> import numpy as np
+        >>> a = np.random.rand(10, 20)
+        >>> bm = BlockMatrix.from_numpy(a)
+
+        Notes
+        -----
+        The ndarray must have two dimensions, each of non-zero size.
+
+        The example above is implemented as:
+
+        >>> a.tofile('local/temp/file') # doctest: +SKIP
+        >>> bm = BlockMatrix.fromfile('file:///local/temp/file', 10, 20) # doctest: +SKIP
+
+        where the local temp file is created on the driver using Python's
+        `tempfile <https://docs.python.org/3/library/tempfile.html>`__
+        library and deleted at exit.
```

Comment:
So, loosely I think of documented things as part of the interface.  We should have the option to change the implementation, and what's more, this is probably more information than the users need.